# Analyzing borrowers’ risk of defaulting

the project is to prepare a report for a bank’s loan division. we need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

the report will be considered when building the **credit score** of a potential customer. The **credit score** is used to evaluate the ability of a potential borrower to repay their loan.

the purpose of the project to test 4 hypotheses about the bank's customers to form recommedation to build a credit scoring model
we will use the data to test the hypotheses about the relationship between a certain customer characteristic and his possible dely in loan pyment 
**Test four hypotheses:**

1.the ability to repay a loan depends on the family's number of kids.<br>
2.marital status affects repaying a loan on time.<br>
3.The level of income predicts a proper loan repayment.<br>
4.loan purposes affect on-time loan repayment.<br>

## loading the data. 

In [1]:
# Loading all the libraries
import pandas as pd

In [2]:
# Load the data
try:
    credit_scoring = pd.read_csv('/datasets/credit_scoring_eng.csv')
except:
    credit_scoring = pd.read_csv(r'C:\Users\Alar\Downloads\credit_scoring_eng.csv')

## Data exploration

**Description of the data**
- `children` - the number of children in the family
- `days_employed` - work experience in days
- `dob_years` - client's age in years
- `education` - client's education
- `education_id` - education identifier
- `family_status` - marital status
- `family_status_id` - marital status identifier
- `gender` - gender of the client
- `income_type` - type of employment
- `debt` - was there any debt on loan repayment
- `total_income` - monthly income
- `purpose` - the purpose of obtaining a loan


In [3]:
#number of rows:
len(credit_scoring) 

21525

In [4]:
#number of columns:
len(credit_scoring.columns)

12

Obtaining the general information about the table:

In [5]:
#general information about the table
credit_scoring.sample(15)

children  days_employed  dob_years            education  education_id  \
15632         0   -1601.586810         47    bachelor's degree             0   
8711          0   -1689.082200         40  secondary education             1   
10510         2    -772.300947         27    bachelor's degree             0   
13578         2    -922.319062         41  secondary education             1   
18359         0   -8285.854949         48  secondary education             1   
14273         0   -1050.845830         33  secondary education             1   
11025         0    -973.896025         46    bachelor's degree             0   
9483          0   -1660.241660         55  secondary education             1   
9712          2   -1014.291387         32  secondary education             1   
7943          0   -3354.454902         47  secondary education             1   
17768         1   -2464.192028         28  secondary education             1   
17665         2    -432.883139         30    bachelor's degree             0   
3243          0  369854.496987         67  secondary education             1   
7204          0            NaN         55  secondary education             1   
13159         2   -6344.469419         43  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
15632            married                 0      M    employee     0   
8711             married                 0      F    employee     0   
10510            married                 0      F    business     0   
13578            married                 0      M    employee     0   
18359           divorced                 3      F    business     0   
14273          unmarried                 4      F    employee     0   
11025            married                 0      F    employee     0   
9483             married                 0      M    employee     1   
9712             married                 0      M    employee     1   
7943             married                 0      F    employee     0   
17768  civil partnership                 1      F    business     0   
17665          unmarried                 4      F    business     0   
3243             married                 0      F     retiree     0   
7204             married                 0      M    employee     0   
13159            married                 0      F    employee     0   

       total_income                                   purpose  
15632     15493.872                  purchase of my own house  
8711      19159.803                                   housing  
10510     36741.225                      university education  
13578     25258.716                                  property  
18359     18706.948                      university education  
14273     20989.782                buy commercial real estate  
11025     24770.492                  purchase of my own house  
9483      18035.779                        housing renovation  
9712      27333.887                         purchase of a car  
7943      20426.354              construction of own property  
17768     37381.146                          wedding ceremony  
17665     64006.056  transactions with commercial real estate  
3243       6703.261                         profile education  
7204            NaN                                  property  
13159      7548.964                   supplementary education

 **Conclusions**
Each row in the table stores data on a client. Some columns have information about the client himself and his family. The rest convey information about the bank account, such as debt column; it shows if there was any debt on loan repayment. And also we can see in purpose column what was the purpose of the loan

It's clear that the data is sufficient to test the hypotheses. However, there are missing values and there is a negative number of days employed

To move forward, we need to preprocess the data.

In [6]:
# Get info on data
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**In columns days_employed and total_income there are missing values, and for both there (21525-19351)=2174 values are missing
We should check if there is any correlation why both columns have the same number of missing values**

In [7]:
# Let's look in the filtered table at the the first column with missing data
credit_scoring[credit_scoring['days_employed'].isnull()].head(20)

children  days_employed  dob_years            education  education_id  \
12          0            NaN         65  secondary education             1   
26          0            NaN         41  secondary education             1   
29          0            NaN         63  secondary education             1   
41          0            NaN         50  secondary education             1   
55          0            NaN         54  secondary education             1   
65          0            NaN         21  secondary education             1   
67          0            NaN         52    bachelor's degree             0   
72          1            NaN         32    bachelor's degree             0   
82          2            NaN         50    bachelor's degree             0   
83          0            NaN         52  secondary education             1   
90          2            NaN         35    bachelor's degree             0   
94          1            NaN         34    bachelor's degree             0   
96          0            NaN         44  SECONDARY EDUCATION             1   
97          0            NaN         47    bachelor's degree             0   
120         0            NaN         46  secondary education             1   
121         0            NaN         29    bachelor's degree             0   
135         0            NaN         27  secondary education             1   
141         0            NaN         39  secondary education             1   
145         0            NaN         62  secondary education             1   
174         0            NaN         55    bachelor's degree             0   

         family_status  family_status_id gender    income_type  debt  \
12   civil partnership                 1      M        retiree     0   
26             married                 0      M  civil servant     0   
29           unmarried                 4      F        retiree     0   
41             married                 0      F  civil servant     0   
55   civil partnership                 1      F        retiree     1   
65           unmarried                 4      M       business     0   
67             married                 0      F        retiree     0   
72             married                 0      M  civil servant     0   
82             married                 0      F       employee     0   
83             married                 0      M       employee     0   
90             married                 0      F       employee     0   
94   civil partnership                 1      F       business     0   
96             married                 0      F       employee     0   
97             married                 0      F       employee     0   
120            married                 0      F       employee     0   
121            married                 0      F       business     0   
135            married                 0      M       business     0   
141  civil partnership                 1      M       employee     0   
145            married                 0      M        retiree     0   
174    widow / widower                 2      F       business     0   

     total_income                                   purpose  
12            NaN                         to have a wedding  
26            NaN                                 education  
29            NaN                    building a real estate  
41            NaN                  second-hand car purchase  
55            NaN                         to have a wedding  
65            NaN  transactions with commercial real estate  
67            NaN       purchase of the house for my family  
72            NaN  transactions with commercial real estate  
82            NaN                                   housing  
83            NaN                                   housing  
90            NaN                      housing transactions  
94            NaN                          having a wedding  
96            NaN               buy residential

In [8]:
credit_scoring[(credit_scoring['days_employed'].isnull()) & (credit_scoring['total_income'].isnull())]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

We can see that NaNs in both columns appear in the same rows, but we must dig deeper into the analysis to find a correlation between missing values in both columns 

missing values in 'total_income' can affect the report for a bank’s loan division. so we will have to:<br>
1.Fill in these missing values with markers<br>
2.Evaluate how much the missing values may affect my computations

In [9]:
credit_scoring.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [10]:
# Let's apply multiple conditions for filtering data and look at the number of rows in the filtered table.

missing_credit_scoring = credit_scoring[credit_scoring.isnull().any(axis=1)]
missing_credit_scoring

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

In [11]:
for i in credit_scoring:
    if credit_scoring[i].isnull().sum()>0:
        print(i)

days_employed
total_income


**Intermediate conclusion**

number of rows in the filtered table match the number of missing values, Now we need to check if the reason there is missing data is because of another due to another specific client characteristic.<br>
 the percentage of the missing values compared to the whole dataset:

In [12]:
percent_missing = credit_scoring.isnull().sum() / len(credit_scoring)
percent_missing

children            0.000000
days_employed       0.100999
dob_years           0.000000
education           0.000000
education_id        0.000000
family_status       0.000000
family_status_id    0.000000
gender              0.000000
income_type         0.000000
debt                0.000000
total_income        0.100999
purpose             0.000000
dtype: float64

 we can see that the missing values is a large piece of data

i want to check which income type has the most missing values in the total_income column to see if there is any relation 

In [13]:
# Let's investigate clients who do not have data on identified characteristic and the column with the missing values
credit_scoring[credit_scoring.total_income.isnull()]['income_type'].value_counts()

employee         1105
business          508
retiree           413
civil servant     147
entrepreneur        1
Name: income_type, dtype: int64

Here we calculate the distribution of the subset of missing values with income_type. And what we get is that employee, business, retiree, civil servant, and entrepreneur have missing values in total_income. We can see below its not for all income type only for few of them.

In [14]:
# Checking distribution
credit_scoring.income_type.value_counts()





employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
unemployed                         2
entrepreneur                       2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64

In [15]:
print((credit_scoring['income_type'].value_counts()/len(credit_scoring))*100)

employee                       51.656214
business                       23.623693
retiree                        17.914053
civil servant                   6.778165
unemployed                      0.009292
entrepreneur                    0.009292
student                         0.004646
paternity / maternity leave     0.004646
Name: income_type, dtype: float64


now we want to check the distribution of family_status with the missing values

In [16]:
credit_scoring[credit_scoring.total_income.isnull()]['family_status'].value_counts()

married              1237
civil partnership     442
unmarried             288
divorced              112
widow / widower        95
Name: family_status, dtype: int64

In [17]:
credit_scoring.family_status.value_counts(normalize=True) 

married              0.575145
civil partnership    0.194053
unmarried            0.130685
divorced             0.055517
widow / widower      0.044599
Name: family_status, dtype: float64

In [18]:
# Checking the distribution in the whole dataset
for i in credit_scoring:
    print(credit_scoring[credit_scoring.total_income.isnull()][i].value_counts()
)
    print('-'*20)



 0     1439
 1      475
 2      204
 3       36
 20       9
 4        7
-1        3
 5        1
Name: children, dtype: int64
--------------------
Series([], Name: days_employed, dtype: int64)
--------------------
34    69
40    66
42    65
31    65
35    64
36    63
47    59
41    59
30    58
28    57
58    56
57    56
54    55
56    54
38    54
52    53
37    53
33    51
39    51
50    51
43    50
45    50
49    50
51    50
29    50
46    48
55    48
48    46
44    44
53    44
60    39
62    38
61    38
32    37
64    37
23    36
27    36
26    35
59    34
63    29
25    23
24    21
65    20
66    20
21    18
22    17
67    16
0     10
68     9
71     5
69     5
20     5
70     3
72     2
19     1
73     1
Name: dob_years, dtype: int64
--------------------
secondary education    1408
bachelor's degree       496
SECONDARY EDUCATION      67
Secondary Education      65
some college             55
Bachelor's Degree        25
BACHELOR'S DEGREE        23
primary education        19
Some Col

**Intermediate conclusion**

The distribution in the original dataset is not similar to the distribution in the filtered table, meaning it could be a pattern between missing values.
To make sure we have investigate our dataset further



In [19]:
# Check for other reasons and patterns that could lead to missing values

missing_credit_scoring_nan_pivot = missing_credit_scoring.pivot_table(index= 'dob_years', columns= 'income_type', values= 'debt', aggfunc='count', margins = True )

In [20]:
missing_credit_scoring_nan_pivot.sort_values(by='All', ascending=False)

income_type  business  civil servant  employee  entrepreneur  retiree   All
dob_years                                                                  
All             508.0          147.0    1105.0           1.0    413.0  2174
34               22.0            4.0      43.0           NaN      NaN    69
40               26.0            9.0      30.0           NaN      1.0    66
42               20.0            1.0      40.0           NaN      4.0    65
31               28.0            7.0      29.0           NaN      1.0    65
35               25.0            2.0      37.0           NaN      NaN    64
36               16.0            9.0      36.0           NaN      2.0    63
47               16.0            5.0      38.0           NaN      NaN    59
41               14.0            2.0      42.0           NaN      1.0    59
30               10.0            4.0      44.0           NaN      NaN    58
28               22.0            4.0      31.0           NaN      NaN    57
58                6.0            3.0      14.0           1.0     32.0    56
57                4.0            2.0      13.0           NaN     37.0    56
54               13.0            2.0      22.0           NaN     18.0    55
38               21.0            3.0      29.0           NaN      1.0    54
56                7.0            4.0      18.0           NaN     25.0    54
37               16.0            2.0      35.0           NaN      NaN    53
52               16.0            5.0      21.0           NaN     11.0    53
33               12.0            7.0      32.0           NaN      NaN    51
39               10.0            6.0      34.0           NaN      1.0    51
50               15.0            4.0      27.0           NaN      5.0    51
49                9.0            3.0      36.0           NaN      2.0    50
45               13.0            2.0      32.0           NaN      3.0    50
29               13.0            6.0      31.0           NaN      NaN    50
51               17.0            3.0      20.0           NaN     10.0    50
43               13.0            4.0      33.0           NaN      NaN    50
55               15.0            3.0      15.0           NaN     15.0    48
46                8.0            5.0      34.0           NaN      1.0    48
48               12.0            7.0      24.0           NaN      3.0    46
53                8.0            2.0      19.0           NaN     15.0    44
44                8.0            3.0      33.0           NaN      NaN    44
60                4.0            2.0       6.0           NaN     27.0    39
61                2.0            NaN       8.0           NaN     28.0    38
62                3.0            1.0       6.0           NaN     28.0    38
32                8.0            4.0      25.0           NaN      NaN    37
64                NaN            3.0       5.0           NaN     29.0    37
27                6.0            3.0      27.0           NaN      NaN    36
23                5.0            1.0      30.0           NaN      NaN    36
26                9.0            2.0      24.0           NaN      NaN    35
59                4.0            1.0       7.0           NaN     22.0    34
63                NaN            1.0      10.0           NaN     18.0    29
25                4.0            4.0      15.0           NaN      NaN    23
24                9.0            1.0      10.0           NaN      1.0    21
65                NaN            NaN       3.0           NaN     17.0    20
66                1.0            NaN       1.0           NaN     18.0    20
21                7.0            1.0      10.0           NaN      NaN    18
22                6.0            NaN      11.0           NaN      NaN    17
67                NaN            NaN       1.0           NaN     15.0    16
0                 2.0            NaN       5.0           NaN      3.0    10
68                NaN            NaN       2.0           NaN      7.0     9
20                1.0            NaN       4.0  

In [21]:
# Checking for other patterns - explain which
missing_credit_scoring_nan_pivot.sort_values(by='employee', ascending=False)

income_type  business  civil servant  employee  entrepreneur  retiree   All
dob_years                                                                  
All             508.0          147.0    1105.0           1.0    413.0  2174
30               10.0            4.0      44.0           NaN      NaN    58
34               22.0            4.0      43.0           NaN      NaN    69
41               14.0            2.0      42.0           NaN      1.0    59
42               20.0            1.0      40.0           NaN      4.0    65
47               16.0            5.0      38.0           NaN      NaN    59
35               25.0            2.0      37.0           NaN      NaN    64
49                9.0            3.0      36.0           NaN      2.0    50
36               16.0            9.0      36.0           NaN      2.0    63
37               16.0            2.0      35.0           NaN      NaN    53
46                8.0            5.0      34.0           NaN      1.0    48
39               10.0            6.0      34.0           NaN      1.0    51
44                8.0            3.0      33.0           NaN      NaN    44
43               13.0            4.0      33.0           NaN      NaN    50
33               12.0            7.0      32.0           NaN      NaN    51
45               13.0            2.0      32.0           NaN      3.0    50
29               13.0            6.0      31.0           NaN      NaN    50
28               22.0            4.0      31.0           NaN      NaN    57
40               26.0            9.0      30.0           NaN      1.0    66
23                5.0            1.0      30.0           NaN      NaN    36
38               21.0            3.0      29.0           NaN      1.0    54
31               28.0            7.0      29.0           NaN      1.0    65
27                6.0            3.0      27.0           NaN      NaN    36
50               15.0            4.0      27.0           NaN      5.0    51
32                8.0            4.0      25.0           NaN      NaN    37
48               12.0            7.0      24.0           NaN      3.0    46
26                9.0            2.0      24.0           NaN      NaN    35
54               13.0            2.0      22.0           NaN     18.0    55
52               16.0            5.0      21.0           NaN     11.0    53
51               17.0            3.0      20.0           NaN     10.0    50
53                8.0            2.0      19.0           NaN     15.0    44
56                7.0            4.0      18.0           NaN     25.0    54
25                4.0            4.0      15.0           NaN      NaN    23
55               15.0            3.0      15.0           NaN     15.0    48
58                6.0            3.0      14.0           1.0     32.0    56
57                4.0            2.0      13.0           NaN     37.0    56
22                6.0            NaN      11.0           NaN      NaN    17
63                NaN            1.0      10.0           NaN     18.0    29
24                9.0            1.0      10.0           NaN      1.0    21
21                7.0            1.0      10.0           NaN      NaN    18
61                2.0            NaN       8.0           NaN     28.0    38
59                4.0            1.0       7.0           NaN     22.0    34
60                4.0            2.0       6.0           NaN     27.0    39
62                3.0            1.0       6.0           NaN     28.0    38
64                NaN            3.0       5.0           NaN     29.0    37
0                 2.0            NaN       5.0           NaN      3.0    10
20                1.0            NaN       4.0           NaN      NaN     5
65                NaN            NaN       3.0           NaN     17.0    20
68                NaN            NaN       2.0           NaN      7.0     9
19                NaN            NaN       1.0           NaN      NaN     1
66                1.0            NaN       1.0  

**Intermediate Conclusions**

we can see the data with missing values according to the type of employment and the age of the client. for example, entrepreneur we have alot of NaNs for most ages this give as better ideas where is our problem

Now i am going to check all the values in columns - if there any duplicate data i need to check how to fix this issue.
Also if there any other issue, like incorrect artifacts. I have to make a decision on how to deal with it.
Then i need to to check the missing values
After the data is clean i can start to work on the data to check the hypothesiss


## Data transformation

Let's go through each column to see what issues we may have in them


Now lets get rid of implicit duplicates in the education column. For example, 'secondary education' be written in different ways. Such errors will also affect the result.

In [22]:
# Let's see all values in education column to check if and what spellings will need to be fixed
credit_scoring.education.unique()

array(["bachelor's degree", 'secondary education', 'Secondary Education',
       'SECONDARY EDUCATION', "BACHELOR'S DEGREE", 'some college',
       'primary education', "Bachelor's Degree", 'SOME COLLEGE',
       'Some College', 'PRIMARY EDUCATION', 'Primary Education',
       'Graduate Degree', 'GRADUATE DEGREE', 'graduate degree'],
      dtype=object)

In [23]:
# Fix the registers if required
credit_scoring['education'] = credit_scoring['education'].str.lower()

In [24]:
# Checking all the values in the column to make sure we fixed them
credit_scoring.education.unique()


array(["bachelor's degree", 'secondary education', 'some college',
       'primary education', 'graduate degree'], dtype=object)

**lets Check the data the `children` column**

In [25]:
# Let's see the distribution of values in the `children` column
credit_scoring['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [26]:
print((credit_scoring['children'].value_counts()/len(credit_scoring))*100)

 0     65.732869
 1     22.383275
 2      9.547038
 3      1.533101
 20     0.353078
-1      0.218351
 4      0.190476
 5      0.041812
Name: children, dtype: float64


We can find the percentage of problematic data is 0.571429; for numbers 20 and -1; the percentage of problematic data is less than 1% . Due to an error the data was transformed wrongly , so we will change -1 to 1 and 20 to 2

In [27]:
# [fix the data based on your decision]
credit_scoring['children']=credit_scoring['children'].replace([-1], value= 1)

In [28]:
credit_scoring['children']=credit_scoring['children'].replace([20], value= 2)

In [29]:
# Checking the `children` column again to make sure it's all fixed

credit_scoring['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

lets Check the data in the `days_employed` column. Firstly we need to think about what kind of issues could there be and what we may want to check and how you will do it.

In [30]:
# Find problematic data in `days_employed`, if they exist, and calculate the percentage
credit_scoring['days_employed'].unique()

array([-8437.67302776, -4024.80375385, -5623.42261023, ...,
       -2113.3468877 , -3112.4817052 , -1984.50758853])

We have negative values, there was probably a problem with the data export<br>
we will use abs() function that Returns the absolute value of a number


In [31]:
# Address the problematic values, if they exist

credit_scoring['days_employed'] = credit_scoring['days_employed'].abs()

In [32]:
# Check the result - make sure it's fixed
credit_scoring['days_employed'].unique()

array([8437.67302776, 4024.80375385, 5623.42261023, ..., 2113.3468877 ,
       3112.4817052 , 1984.50758853])

In [33]:
credit_scoring['days_employed'].max()

401755.40047533

i can see that the number of days_employes does not make sense -I set a number of days of 50 years as the limit

In [34]:
limit = 365 * 50

In [35]:
credit_scoring['days_employed'][credit_scoring['days_employed'] > limit ].count()

3446

we have 3446 values that are weird values i think that the values in the column is not vaild/ normal

Let's now look at the client's age and whether there are any issues there. what cannot be someone's age.

In [36]:
# Check the `dob_years` for suspicious values and count the percentage
print((credit_scoring['dob_years'].value_counts()/len(credit_scoring))*100)


35    2.866434
40    2.829268
41    2.819977
34    2.801394
38    2.778165
42    2.773519
33    2.699187
39    2.662021
31    2.601626
36    2.578397
44    2.541231
29    2.531940
30    2.508711
48    2.499419
37    2.494774
50    2.387921
43    2.383275
32    2.369338
49    2.360046
28    2.336818
45    2.308943
27    2.290360
56    2.262485
52    2.248548
47    2.229965
54    2.225319
46    2.206736
58    2.141696
57    2.137050
53    2.132404
51    2.081301
59    2.062718
55    2.058072
26    1.895470
60    1.751452
25    1.658537
61    1.649245
62    1.635308
63    1.249710
64    1.231127
24    1.226481
23    1.180023
65    0.901278
22    0.850174
66    0.850174
67    0.775842
21    0.515679
0     0.469222
68    0.459930
69    0.394890
70    0.301974
71    0.269454
20    0.236934
72    0.153310
19    0.065041
73    0.037166
74    0.027875
75    0.004646
Name: dob_years, dtype: float64


we have '0' in dob_years, as we saw in the percentage 0.469222% it less than 1% we can sort values to see if we can find sequen between the age and then replace  the 0 with the missing value

In [37]:
credit_scoring['dob_years'].sort_values().unique()

array([ 0, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
       52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68,
       69, 70, 71, 72, 73, 74, 75], dtype=int64)

We cannot determine which age is missing because from 19 to 75 we have all the values so we had to delete the rows with value 0. we cant fill 0 with NaN or ' ' because we want the type of the column as an integer 

In [38]:
# Address the issues in the `dob_years` column, if they exist
credit_scoring = credit_scoring.drop(credit_scoring.index[credit_scoring['dob_years'] == 0])

In [39]:
# Check the result - make sure it's fixed
credit_scoring['dob_years'].unique()

array([42, 36, 33, 32, 53, 27, 43, 50, 35, 41, 40, 65, 54, 56, 26, 48, 24,
       21, 57, 67, 28, 63, 62, 47, 34, 68, 25, 31, 30, 20, 49, 37, 45, 61,
       64, 44, 52, 46, 23, 38, 39, 51, 59, 29, 60, 55, 58, 71, 22, 73, 66,
       69, 19, 72, 70, 74, 75], dtype=int64)

Now let's check the `family_status` column.  what kind of values there are and what problems i need to address.

In [40]:
# Let's see the values for the column
credit_scoring['family_status'].sort_values().unique()


array(['civil partnership', 'divorced', 'married', 'unmarried',
       'widow / widower'], dtype=object)

**I dont see a problem with the values**

**Now let's check the `gender` column.  what kind of values there are and what problems i need to address**

In [41]:
# Let's see the values in the column
credit_scoring['gender'].value_counts()

F      14164
M       7259
XNA        1
Name: gender, dtype: int64

we can see there is 'XNA' value and it's only appear in one row ,i chose to delete the row 

In [42]:
# Address the problematic values, if they exist
credit_scoring = credit_scoring.drop(credit_scoring.index[credit_scoring['gender'] == 'XNA'])

In [43]:
# Check the result - make sure it's fixed
credit_scoring['gender'].value_counts()


F    14164
M     7259
Name: gender, dtype: int64

Now let's check the `income_type` column. what kind of values there are and what problems i need to address

In [44]:
# Let's see the values in the column
credit_scoring['income_type'].value_counts()

employee                       11064
business                        5064
retiree                         3836
civil servant                   1453
unemployed                         2
entrepreneur                       2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64


there is no problematic values 

Now let's see if we have any duplicates in our data

In [45]:
# Checking duplicates
credit_scoring.duplicated().sum()


71

In [46]:
# Address the duplicates, if they exist
credit_scoring = credit_scoring.drop_duplicates()

In [47]:
credit_scoring = credit_scoring.reset_index(drop=True)

In [48]:
# Last check whether we have any duplicates
credit_scoring.duplicated().sum()

0

i chose to drop all duplicate data 

In [49]:
# Check the size of the dataset that you now have after your first manipulations with it
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21352 entries, 0 to 21351
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21352 non-null  int64  
 1   days_employed     19259 non-null  float64
 2   dob_years         21352 non-null  int64  
 3   education         21352 non-null  object 
 4   education_id      21352 non-null  int64  
 5   family_status     21352 non-null  object 
 6   family_status_id  21352 non-null  int64  
 7   gender            21352 non-null  object 
 8   income_type       21352 non-null  object 
 9   debt              21352 non-null  int64  
 10  total_income      19259 non-null  float64
 11  purpose           21352 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [50]:
(21352-19259)/21352

0.09802360434619707

The number of rows is less than it was because we removed the duplicate values and the columns that contained missing data also became less accordingly. in addition the datatype of the columns changed for some of columns

now we need to replace the name that could be names written incorrectly or alternative names of the same purpose. 

In [51]:
credit_scoring['purpose'].sort_values().unique()

array(['building a property', 'building a real estate',
       'buy commercial real estate', 'buy real estate',
       'buy residential real estate', 'buying a second-hand car',
       'buying my own car', 'buying property for renting out', 'car',
       'car purchase', 'cars', 'construction of own property',
       'education', 'getting an education', 'getting higher education',
       'going to university', 'having a wedding', 'housing',
       'housing renovation', 'housing transactions', 'profile education',
       'property', 'purchase of a car', 'purchase of my own house',
       'purchase of the house', 'purchase of the house for my family',
       'real estate transactions', 'second-hand car purchase',
       'supplementary education', 'to become educated', 'to buy a car',
       'to get a supplementary education', 'to have a wedding',
       'to own a car', 'transactions with commercial real estate',
       'transactions with my real estate', 'university education',
       'we

In [52]:
credit_scoring['purpose'] = credit_scoring['purpose'].fillna('')

for i in range(0,len(credit_scoring)):
    if "car" in credit_scoring.loc[i, "purpose"]:
        credit_scoring.loc[i, "purpose"] = "Car"
    elif "wedding" in credit_scoring.loc[i, "purpose"]:
        credit_scoring.loc[i, "purpose"] = "Wedding"
    elif "educat" in credit_scoring.loc[i, "purpose"]:
        credit_scoring.loc[i, "purpose"] = "education"
    elif "university" in credit_scoring.loc[i, "purpose"]:
        credit_scoring.loc[i, "purpose"] = "education"
    elif "real estate" or "property"  in credit_scoring.loc[i, "purpose"]:
        credit_scoring.loc[i, "purpose"] = "real estate"
    else:    
        print(credit_scoring.loc[i, "purpose"])
        

In [53]:
credit_scoring['purpose'].sort_values().unique()

array(['Car', 'Wedding', 'education', 'real estate'], dtype=object)

We detected three issues with the data:

- Obvious and implicit duplicates
- We handled problematic data
The headers have been cleaned up to make processing the table simpler.

# Working with missing values

i choose to work with income_type and total_type Because of the relationship between the two columns if we could find a pattern for each income type we could fill the missing values in total income

In [54]:
# Find the dictionaries
credit_scoring['total_income_1']=credit_scoring['total_income']


In [55]:
#group the data by income_type and calculate mean total_income for each job
grouped_data=credit_scoring.groupby(['income_type'])['total_income'].mean()
grouped_data

income_type
business                       32397.307219
civil servant                  27361.316126
employee                       25824.679592
entrepreneur                   79866.103000
paternity / maternity leave     8612.661000
retiree                        21939.310393
student                        15712.260000
unemployed                     21014.360500
Name: total_income, dtype: float64

In [56]:
#pass it to the dictionary
grouped_data_dict=pd.Series(grouped_data).to_dict()
grouped_data_dict

{'business': 32397.30721939447,
 'civil servant': 27361.31612557427,
 'employee': 25824.679592232038,
 'entrepreneur': 79866.103,
 'paternity / maternity leave': 8612.661,
 'retiree': 21939.310392586107,
 'student': 15712.26,
 'unemployed': 21014.360500000003}

In [57]:
grouped_data_dict['student']

15712.26

In [58]:
grouped_data_dict.values()

dict_values([32397.30721939447, 27361.31612557427, 25824.679592232038, 79866.103, 8612.661, 21939.310392586107, 15712.26, 21014.360500000003])

In [59]:
grouped_data_dict.keys()

dict_keys(['business', 'civil servant', 'employee', 'entrepreneur', 'paternity / maternity leave', 'retiree', 'student', 'unemployed'])

In [60]:
grouped_data_dict

{'business': 32397.30721939447,
 'civil servant': 27361.31612557427,
 'employee': 25824.679592232038,
 'entrepreneur': 79866.103,
 'paternity / maternity leave': 8612.661,
 'retiree': 21939.310392586107,
 'student': 15712.26,
 'unemployed': 21014.360500000003}

In [61]:
for key,value in grouped_data_dict.items():
        print(f'{key} earns {value}')

business earns 32397.30721939447
civil servant earns 27361.31612557427
employee earns 25824.679592232038
entrepreneur earns 79866.103
paternity / maternity leave earns 8612.661
retiree earns 21939.310392586107
student earns 15712.26
unemployed earns 21014.360500000003


In [62]:
 for key,value in grouped_data_dict.items():
        print('{} earns {}'.format(key,value))

business earns 32397.30721939447
civil servant earns 27361.31612557427
employee earns 25824.679592232038
entrepreneur earns 79866.103
paternity / maternity leave earns 8612.661
retiree earns 21939.310392586107
student earns 15712.26
unemployed earns 21014.360500000003


In [63]:
credit_scoring[credit_scoring['total_income_1'].isna()]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21316         2            NaN         47  secondary education             1   
21322         1            NaN         50  secondary education             1   
21324         0            NaN         48    bachelor's degree             0   
21329         1            NaN         42  secondary education             1   
21337         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21316            married                 0      M       business     0   
21322  civil partnership                 1      F       employee     0   
21324            married                 0      F       business     0   
21329            married                 0      F       employee     0   
21337            married                 0      F       employee     0   

       total_income      purpose  total_income_1  
12              NaN      Wedding             NaN  
26              NaN    education             NaN  
29              NaN  real estate             NaN  
41              NaN          Car             NaN  
55              NaN      Wedding             NaN  
...             ...          ...             ...  
21316           NaN          Car             NaN  
21322           NaN      Wedding             NaN  
21324           NaN  real estate             NaN  
21329           NaN  real estate             NaN  
21337           NaN          Car             NaN  

[2093 rows x 13 columns]

We will try to replace the missing values in example column by using fillna() and map() with the dictionary we've creaed.
It will try to replace NaN in days_employed_1 but looking at column income_type and treating income_type as a key to the dictionary grouped_data_dict that has our mean total_income_1

In [64]:
credit_scoring['total_income_1']=credit_scoring['total_income_1'].fillna(credit_scoring.income_type.map(grouped_data_dict))

In [65]:
credit_scoring.loc[12]

children                              0
days_employed                       NaN
dob_years                            65
education           secondary education
education_id                          1
family_status         civil partnership
family_status_id                      1
gender                                M
income_type                     retiree
debt                                  0
total_income                        NaN
purpose                         Wedding
total_income_1             21939.310393
Name: 12, dtype: object

We see that our total_income_1 has been replaced with 21939.31 which is the mean total_income for retiree

In [66]:
grouped_data_dict['retiree']

21939.310392586107

### Restoring missing values in `total_income`

i will Start with addressing total income missing values. for this i need to Create age category for clients. Create a new column with the age category. This strategy can help with calculating values for the total income.


In [67]:
#let's write a function that calculates the age category
def age_category(row):
    age = row['dob_years']

    if age < 25:
        return '<25 '
    elif 26 < age < 45:
        return '26 - 45 '
    elif 46 < age < 65 :
        return '46 - 65'
    else:
        return ' 66<'


In [68]:
# Test if the function works
row_values = [30]
row_columns = ['dob_years']
row = pd.Series(data=row_values, index=row_columns)
print(age_category(row)) 


26 - 45 


In [69]:
# Creating new column based on function
credit_scoring['age_group']= credit_scoring.apply(age_category, axis=1)

In [70]:
# Checking how values in the new column

print(credit_scoring['age_group'].value_counts()) 

26 - 45     10067
46 - 65      7783
 66<         2628
<25           874
Name: age_group, dtype: int64


i need to think about the factors on which income usually depends. Eventually, i need to find out whether i should use mean or median values for replacing missing values. To make this decision i need probably to look at the distribution of the factors i identified as impacting one's income.

wen need to Create a table that only has data without missing values. This data will be used to restore the missing values.

In [71]:
# Create a table without missing values and print a few of its rows to make sure it looks fine
credit_scoring1 = credit_scoring.dropna()
credit_scoring1.head(30)

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42    bachelor's degree             0   
1          1    4024.803754         36  secondary education             1   
2          0    5623.422610         33  secondary education             1   
3          3    4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0     926.185831         27    bachelor's degree             0   
6          0    2879.202052         43    bachelor's degree             0   
7          0     152.779569         50  secondary education             1   
8          2    6929.865299         35    bachelor's degree             0   
9          0    2188.756445         41  secondary education             1   
10         2    4171.483647         36    bachelor's degree             0   
11         0     792.701887         40  secondary education             1   
13         0    1846.641941         54         some college             2   
14         0    1844.956182         56    bachelor's degree             0   
15         1     972.364419         26  secondary education             1   
16         0    1719.934226         35  secondary education             1   
17         0    2369.999720         33    bachelor's degree             0   
18         0  400281.136913         53  secondary education             1   
19         0   10038.818549         48  secondary education             1   
20         1    1311.604166         36  secondary education             1   
21         1     253.685166         33  secondary education             1   
22         1    1766.644138         24  secondary education             1   
23         0     272.981385         21    bachelor's degree             0   
24         1  338551.952911         57  secondary education             1   
25         0  363548.489348         67  secondary education             1   
27         0     529.191635         28    bachelor's degree             0   
28         1     717.274324         26    bachelor's degree             0   
30         1  335581.668515         62  secondary education             1   
31         0    1682.083438         47    primary education             3   
32         0    4649.910832         34  secondary education             1   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   
15            married                 0      F    employee     0   
16            married                 0      F    employee     0   
17  civil partnership                 1      M    employee     0   
18    widow / widower                 2      F     retiree     0   
19           divorced                 3      F    employee     0   
20            married                 0      M    employee     0   
21  civil partnership                 1      F    employee     0   
22  civil partnership                 1      F    employee     0   
23  civil partnership                 1      M    em

In [72]:
# Look at the mean values for income based on your identified factors
credit_scoring1.groupby(['income_type'])['total_income'].mean()


income_type
business                       32397.307219
civil servant                  27361.316126
employee                       25824.679592
entrepreneur                   79866.103000
paternity / maternity leave     8612.661000
retiree                        21939.310393
student                        15712.260000
unemployed                     21014.360500
Name: total_income, dtype: float64

In [73]:
# Look at the median values for income based on your identified factors
credit_scoring1.groupby(['income_type'])['total_income'].median()

income_type
business                       27563.0285
civil servant                  24083.5065
employee                       22815.1035
entrepreneur                   79866.1030
paternity / maternity leave     8612.6610
retiree                        18969.1490
student                        15712.2600
unemployed                     21014.3605
Name: total_income, dtype: float64

In [74]:
credit_scoring.describe()[['total_income']]

total_income
count   19259.000000
mean    26794.133121
std     16494.347322
min      3306.762000
25%     16494.864000
50%     23200.877000
75%     32536.052000
max    362496.645000

In [75]:
credit_scoring1.pivot_table(index= 'income_type', columns='age_group', values= 'total_income', aggfunc= 'mean')

age_group                             66<      26 - 45        46 - 65  \
income_type                                                             
business                     31779.479363  33044.784623  32633.628376   
civil servant                26677.659762  27625.755580  27555.918294   
employee                     25077.396923  26265.957415  26005.201458   
entrepreneur                          NaN  79866.103000           NaN   
paternity / maternity leave           NaN   8612.661000           NaN   
retiree                      20927.585055  22566.785365  22204.346500   
student                               NaN           NaN           NaN   
unemployed                   32435.602000   9593.119000           NaN   

age_group                            <25   
income_type                                
business                     25446.057616  
civil servant                23548.822930  
employee                     21288.654010  
entrepreneur                          NaN  
paternity / maternity leave           NaN  
retiree                      14298.976000  
student                      15712.260000  
unemployed                            NaN

In [76]:
credit_scoring1.pivot_table(index='income_type', columns='education', values='total_income', aggfunc='median')

education                    bachelor's degree  graduate degree  \
income_type                                                       
business                            32265.7180              NaN   
civil servant                       27601.7775        17822.757   
employee                            26535.3245        31771.321   
entrepreneur                        79866.1030              NaN   
paternity / maternity leave                NaN              NaN   
retiree                             23078.5230        28334.215   
student                             15712.2600              NaN   
unemployed                          32435.6020              NaN   

education                    primary education  secondary education  \
income_type                                                           
business                             21887.825           25446.2700   
civil servant                        23734.287           21912.6450   
employee                             20159.186           21853.6300   
entrepreneur                               NaN                  NaN   
paternity / maternity leave                NaN            8612.6610   
retiree                              16415.785           18382.4045   
student                                    NaN                  NaN   
unemployed                                 NaN            9593.1190   

education                    some college  
income_type                                
business                        28688.018  
civil servant                   25694.775  
employee                        24209.430  
entrepreneur                          NaN  
paternity / maternity leave           NaN  
retiree                         19221.903  
student                               NaN  
unemployed                            NaN

In [77]:
credit_scoring1.pivot_table(index=('income_type','age_group'), columns='education', values='total_income', aggfunc='median')

education                              bachelor's degree  graduate degree  \
income_type                 age_group                                       
business                     66<              30331.1440              NaN   
                            26 - 45           32671.8580              NaN   
                            46 - 65           34651.0390              NaN   
                            <25               24214.3790              NaN   
civil servant                66<              23624.0095              NaN   
                            26 - 45           27689.2105        17822.757   
                            46 - 65           29612.8500              NaN   
                            <25               24617.5440              NaN   
employee                     66<              24405.3370        31771.321   
                            26 - 45           26884.3335        18551.846   
                            46 - 65           27694.4100        42945.794   
                            <25               21535.1395              NaN   
entrepreneur                26 - 45           79866.1030              NaN   
paternity / maternity leave 26 - 45                  NaN              NaN   
retiree                      66<              21462.4190        15800.399   
                            26 - 45           22993.1380              NaN   
                            46 - 65           23918.5100        40868.031   
                            <25                      NaN              NaN   
student                     <25               15712.2600              NaN   
unemployed                   66<              32435.6020              NaN   
                            26 - 45                  NaN              NaN   

education                              primary education  secondary education  \
income_type                 age_group                                           
business                     66<              27929.0980           26493.4095   
                            26 - 45           21441.9210           25677.3190   
                            46 - 65           23066.1310           25329.9340   
                            <25               27119.0240           21724.3930   
civil servant                66<                     NaN           23512.0360   
                            26 - 45           30545.9490           22054.6790   
                            46 - 65           16922.6250           21417.6660   
                            <25                      NaN           20428.4720   
employee                     66<              28124.9560           21916.2890   
                            26 - 45           19542.3265           22097.1900   
                            46 - 65           20896.0125           21702.4710   
                            <25               24438.8615           18300.5890   
entrepreneur                26 - 45                  NaN                  NaN   
paternity / maternity leave 26 - 45                  NaN            8612.6610   
retiree                      66<              15013.5050           17440.9280   
                            26 - 45           17973.2670           18495.7740   
                            46 - 65           17657.4995           18693.7630   
                            <25                      NaN           14298.9760   
student                     <25                      NaN                  NaN   
unemployed                   66<                     NaN                  NaN   
                            26 - 45                  NaN            9593.1190   

education                              some college  
income_type                 age_group                
business                     66<         29534.1630  
                            26 - 45      29814.5275  
                            46 - 65      37146.5350  
                            <25          20964.2540  
civil servant                66<        

i choose multiple factors. i need to make sure and consider different aspects age and education can effect client income



for total income i chose to use median because the data that i working with has serious outliers. The mean is sensitive to outliers, but the fact the mean is so small compared to the max value indicates the max value is an outlier

In [78]:
#  Write a function that we will use for filling in missing values
credit_scoring['total_income'] = pd.to_numeric(credit_scoring['total_income'], errors='coerce')
credit_scoring['total_income'] = credit_scoring['total_income'].fillna(credit_scoring1.pivot_table)       
        

In [79]:
# Check if it works
credit_scoring.isnull().sum()

children               0
days_employed       2093
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
total_income_1         0
age_group              0
dtype: int64

i didnt delete the rows because i want to know if there is any other way to deal filling in missing values

In [80]:
# Checking the number of entries in the columns
print(len(credit_scoring))
print(len(credit_scoring.total_income))

21352
21352


###  Restoring values in `days_employed`

i need to think about the parameters that may help me to restore the missing values in this column. Eventually,i want to find out whether i should use mean or median values for replacing missing values.

In [81]:
# Distribution of `days_employed` medians based on your identified parameters
credit_scoring1.groupby(['dob_years'])['days_employed'].median()

dob_years
19       724.492610
20       674.838979
21       618.733817
22       703.310078
23       690.204208
24       942.390603
25       919.199388
26      1083.658132
27      1166.212160
28      1141.705450
29      1315.453550
30      1420.586863
31      1308.901135
32      1446.622542
33      1426.415003
34      1615.910188
35      1613.494818
36      1799.520465
37      1816.713382
38      1817.194286
39      1891.388566
40      1728.936706
41      1864.657692
42      2253.981630
43      1869.019899
44      2084.330015
45      2254.431130
46      2100.473217
47      2203.078545
48      2429.674518
49      2560.317106
50      2626.986257
51      2846.080700
52      3395.639568
53      3650.007523
54      4026.541145
55      7043.471139
56      6850.179463
57    333953.491811
58    331245.976666
59    344372.337365
60    349414.537285
61    343984.931910
62    354981.718187
63    358482.295938
64    355781.562599
65    357103.967291
66    363827.438951
67    357222.855215
68    3652

In [82]:
# Distribution of `days_employed` means based on your identified parameters
credit_scoring1.groupby(['dob_years'])['days_employed'].mean()

dob_years
19       633.678086
20       684.944308
21       709.440930
22      2793.321854
23       827.309437
24      1020.900547
25      1088.406453
26      3214.617245
27      3725.820728
28      2180.056457
29      1553.823200
30      1696.039355
31      2331.246721
32      3877.197466
33      3241.413096
34      4141.242883
35      2755.226484
36      4496.534938
37      5879.263440
38      7641.485465
39      5277.456460
40      7076.521918
41      6400.926359
42      8885.465110
43      9418.428468
44      9952.060042
45     11748.039481
46     14121.191502
47     14223.685900
48     18149.683912
49     26679.790846
50     50493.031018
51     68425.772762
52     82803.402429
53     94073.249426
54    126757.781831
55    152264.946612
56    157417.505013
57    193689.068297
58    188571.489818
59    227496.974028
60    262925.277586
61    246731.048849
62    274381.283863
63    293383.721011
64    287670.475737
65    287448.538552
66    313613.718234
67    317629.788586
68    3268

In [83]:
credit_scoring.describe()[['days_employed']]

days_employed
count   19259.000000
mean    66905.789237
std    139025.028132
min        24.141633
25%       926.823974
50%      2197.320350
75%      5540.399763
max    401755.400475

for days_employed i chose to use median because the data that i working with has serious outliers.
For example, the max is 401755.400475 while its mean is 66905.789237. The mean is sensitive to outliers, but the fact the mean is so small compared to the max value indicates the max value is an outlier

In [84]:
# Let's write a function that calculates means or medians (depending on your decision) based on your identified parameter
credit_scoring['days_employed'] = pd.to_numeric(credit_scoring['days_employed'], errors='coerce')
credit_scoring['days_employed'] = credit_scoring['days_employed'].fillna(credit_scoring.groupby('dob_years')['days_employed'].transform('median'))       

In [85]:
# Check that the function works

credit_scoring.isnull().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
total_income_1      0
age_group           0
dtype: int64

In [86]:
# Check the entries in all columns - make sure we fixed all missing values
print(len(credit_scoring))
print(len(credit_scoring.days_employed))

21352
21352


## Categorization of data

[To answer the questions and test the hypotheses, you will want to work with categorized data. Look at the questions that were posed to you and that you should answer. Think about which of the data will need to be categorized to answer these questions. Below you will find a template through which you can work your way when categorizing data. The first step-by-step processing covers the text data; the second one addresses the numerical data that needs to be categorized. You can use both or none of the suggested instructions - it's up to you.]

[Despite of how you decide to address the categorization, make sure to provide clear explanation of why you made your decision. Remember: this is your work and you make all decisions in it.]


In [87]:
# Print the values for your selected data for categorization
credit_scoring[['children', 'debt']]

children  debt
0             1     0
1             1     0
2             0     0
3             3     0
4             0     0
...         ...   ...
21347         1     0
21348         0     0
21349         1     1
21350         3     1
21351         2     0

[21352 rows x 2 columns]

[Let's check unique values]

In [88]:
# Check the unique values
credit_scoring['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

the main groups identify based on children numbers in the family 

In [89]:
# Let's write a function to categorize the data based on common topics
def family_categories(row):
    member = row['children']

    if member <= 1:
        return 'small family'
    elif member <= 3:
        return 'medium family'
    else:
        return 'big family'

In [90]:
# Create a column with the categories and count the values for them
credit_scoring['family_group']= credit_scoring.apply(family_categories, axis=1)

[If you decide to categorize the numerical data, you'll need to come up with the categories for it too.]

In [91]:
# Looking through all the numerical data in your selected column for categorization
credit_scoring[['family_group','debt','children']]


family_group  debt  children
0       small family     0         1
1       small family     0         1
2       small family     0         0
3      medium family     0         3
4       small family     0         0
...              ...   ...       ...
21347   small family     0         1
21348   small family     0         0
21349   small family     1         1
21350  medium family     1         3
21351  medium family     0         2

[21352 rows x 3 columns]

In [92]:
# Getting summary statistics for the column
credit_scoring.groupby('family_group').describe()

children                                               \
                 count      mean       std  min  25%  50%  75%  max   
family_group                                                          
big family        50.0  4.180000  0.388088  4.0  4.0  4.0  4.0  5.0   
medium family   2442.0  2.134316  0.341061  2.0  2.0  2.0  2.0  3.0   
small family   18860.0  0.256575  0.436754  0.0  0.0  0.0  1.0  1.0   

              days_employed                ... debt      total_income_1  \
                      count          mean  ...  75%  max          count   
family_group                               ...                            
big family             50.0   9988.782494  ...  0.0  1.0           50.0   
medium family        2442.0   6962.903403  ...  0.0  1.0         2442.0   
small family        18860.0  74587.961909  ...  0.0  1.0        18860.0   

                                                                               \
                       mean           std       min         25%           50%   
family_group                                                                    
big family     27144.246647  12002.022583  7803.663  20210.4765  25824.679592   
medium family  27698.583104  16471.645199  4494.861  17382.6560  25176.448500   
small family   26678.187189  15611.610769  3306.762  17191.1905  24172.269500   

                                         
                        75%         max  
family_group                             
big family     29396.319750   79094.031  
medium family  32397.307219  176552.869  
small family   32384.018250  362496.645  

[3 rows x 56 columns]

The most common family group in our database is 'small family' which they have 0-1 kid - number of families = 18860
the second common family group is 'medium family' - the have 2-3 kids and number of families = 2442.
as we can see there is few who have more then 3 kids in there family - there only 50 families 

In [93]:
credit_scoring['total_income'] = pd.to_numeric(credit_scoring['total_income'],errors='coerce')

In [94]:
# Creating function for categorizing into different numerical groups based on ranges
def salary_categories(row):
    salary = row['total_income']

    if (salary > 0) and (salary <= 17000): 
        return 'low'
    if (salary > 17000) and (salary <= 40000): 
        return 'average'
    if (salary > 40000) and (salary <= 80000):
        return 'above average'
    if (salary > 80000) and (salary <= 200000):
        return 'high'
    if salary > 200000:
        return 'very high'


In [95]:
# Creating column with categories
credit_scoring['salary_group']= credit_scoring.apply(salary_categories, axis=1)

In [96]:
# Count each categories values to see the distribution
credit_scoring[['salary_group', 'total_income','income_type']]

salary_group  total_income income_type
0      above average     40620.102    employee
1            average     17932.802    employee
2            average     23341.752    employee
3      above average     42820.568    employee
4            average     25378.572     retiree
...              ...           ...         ...
21347        average     35966.698    business
21348        average     24959.969     retiree
21349            low     14347.610    employee
21350        average     39054.888    employee
21351            low     13127.587    employee

[21352 rows x 3 columns]

## Checking the Hypotheses


**Is there a correlation between having children and paying back on time?**

In [97]:
# Check the children data and paying back on time

pivot_table_children = credit_scoring.pivot_table(index='children', columns='debt', values='days_employed', aggfunc='count')


In [98]:
pivot_table_children

debt            0       1
children                 
0         12963.0  1058.0
1          4397.0   442.0
2          1912.0   202.0
3           301.0    27.0
4            37.0     4.0
5             9.0     NaN

In [99]:
# Calculating default-rate based on the number of children
pivot_table_children['rate_1'] =pivot_table_children[1]/(pivot_table_children[1]+pivot_table_children[0])*100

In [100]:
pivot_table_children['count'] =pivot_table_children[1]+pivot_table_children[1]

In [101]:
pivot_table_children.sort_values(by='rate_1',ascending=True)

debt            0       1    rate_1   count
children                                   
0         12963.0  1058.0  7.545824  2116.0
3           301.0    27.0  8.231707    54.0
1          4397.0   442.0  9.134119   884.0
2          1912.0   202.0  9.555345   404.0
4            37.0     4.0  9.756098     8.0
5             9.0     NaN       NaN     NaN

**Conclusion**
the rate for client without kids is higher than client with kids, but number of kids doesnt afftect the rate.

**Is there a correlation between family status and paying back on time?**

In [102]:
# Check the family status data and paying back on time
pivot_table_family_status = credit_scoring.pivot_table(index='family_status', columns='debt', values='days_employed', aggfunc='count')


In [103]:
# Calculating default-rate based on family status
pivot_table_family_status['rate_1'] =pivot_table_family_status[1]/(pivot_table_family_status[1]+pivot_table_family_status[0])*100

In [104]:
pivot_table_family_status.sort_values(by='rate_1',ascending=True)

debt                   0    1    rate_1
family_status                          
widow / widower      892   62  6.498952
divorced            1100   85  7.172996
married            11363  927  7.542718
civil partnership   3743  386  9.348511
unmarried           2521  273  9.770938

**Conclusion**
we can see that under the unmarried category there is the highest rate of debt on loan repayment



**Is there a correlation between income level and paying back on time?**

In [105]:
# Check the income level data and paying back on time
pivot_table_income_level = credit_scoring.pivot_table(index='salary_group', columns='debt', values='days_employed', aggfunc='count')

In [106]:
# Calculating default-rate based on income level
pivot_table_income_level['rate_1'] =pivot_table_income_level[1]/(pivot_table_income_level[1]+pivot_table_income_level[0])*100

In [107]:
pivot_table_income_level.sort_values(by='rate_1',ascending=True)

debt               0    1    rate_1
salary_group                       
high             198   13  6.161137
above average   2398  179  6.946061
low             4782  408  7.861272
average        10308  962  8.535936
very high         10    1  9.090909

**Conclusion**
we can see clients with 'above average' income have a debt. in other hand, client with very high income less appears under debt. 

**How does credit purpose affect the default rate?**

In [108]:
# Check the percentages for default rate for each credit purpose and analyze them
pivot_table_purpose = credit_scoring.pivot_table(index='purpose', columns='debt', values='days_employed', aggfunc='count')


In [109]:
pivot_table_purpose['rate_1'] =pivot_table_purpose[1]/(pivot_table_purpose[1]+pivot_table_purpose[0])*100

In [110]:
pivot_table_purpose.sort_values(by='rate_1',ascending=True)

debt            0    1    rate_1
purpose                         
real estate  9984  779  7.237759
Wedding      2126  184  7.965368
education    3625  370  9.261577
Car          3884  400  9.337068

**Conclusion**
We can see that under education and car we have the highest rate.


# General Conclusion 

missing values in columns days_employed and total income are random, we need to check with the team who are responsible for collecting data.
duplicate values seems to be occuard due to tecnical issues, we should recommend to the bank to store information according to customer id.


Findings 
We have tested the following four hypotheses:

There is no correlation between having children and paying loans back on time
in the same way, the small family has the most Inability to repay the loan, and big families Are rated with strong loan repayment capacity
The first hypothesis is not accepted.

Based on the family status we can see that under the unmarried category there is the highest rate of debt on loan repayment So we can not accept this hypothesis.

we can see clients with 'above average' income have a debt. The third hypothesis is rejected. 

Loan for car or education is the most loan types that customers have had trouble repaying.
The last hypothesis is accepted.